### DISCUSSION TITLE
Add attributes to logs

### DISCUSSION THREAD

**Discussion by Flooze-IT** [open-telemetry/opentelemetry-java-instrumentation#7303] [Q&A] [answered]
Hi everyone,

I'm experimenting with the logging SDK, however I'm getting stuck when I need to pass particular attributes to my logs. I can't find how to do it in the documentation.

Here is my configuration:

``` java
@Bean
public SdkLoggerProvider sdkLoggerProvider(){
    Resource resource = Resource.getDefault()
            .merge(Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, "Admin Service")));

    LogRecordExporter logRecordExporter = OtlpHttpLogRecordExporter.builder()
            .setEndpoint("https://my-open-telemetry-endpoint.com"+"/v1/logs")
            .build();

    SdkLoggerProvider sdkLoggerProvider = SdkLoggerProvider.builder()
            .addLogRecordProcessor(BatchLogRecordProcessor.builder(logRecordExporter).build())
            .setResource(resource)
            .build();

    GlobalLoggerProvider.set(sdkLoggerProvider);

    return sdkLoggerProvider;
}
```

``` java
@Bean
public OpenTelemetry openTelemetry(SdkLoggerProvider sdkLoggerProvider) throws IOException {
    Resource resource = Resource.getDefault()
            .merge(Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, "Admin Service")));

    // trace config
    SpanExporter spanExporter = OtlpHttpSpanExporter.builder()
            .setEndpoint("https://my-open-telemetry-endpoint.com"+"/v1/traces")
            .build();

    SdkTracerProvider sdkTracerProvider = SdkTracerProvider.builder()
            .addSpanProcessor(BatchSpanProcessor.builder(spanExporter).build())
            .setResource(resource)
            .build();

    OpenTelemetry openTelemetry = OpenTelemetrySdk.builder()
            .setTracerProvider(sdkTracerProvider)
            .setLoggerProvider(sdkLoggerProvider)
            .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
            .buildAndRegisterGlobal();

    return openTelemetry;
}
```

I use the classic:
``` java
log.error("My message")
``` 
and that work properly but how to add specific attributes ?

Thanks

**Comment by saytoshivam** [comment]
I think u can use Attribute builder to add attributes.
Here is the sample code from where you can have some idea.

`import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.common.AttributesBuilder;
import io.opentelemetry.sdk.resources.Resource;
import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;`


`AttributesBuilder attrBuilders = Attributes.builder();
  attrBuilders.put(
            ResourceAttributes.CLOUD_PLATFORM,
            ResourceAttributes.CloudPlatformValues.GCP_COMPUTE_ENGINE);
Resource.create(attrBuilders.build(), ResourceAttributes.SCHEMA_URL);
`

perhaps it may help you somehow.

**Reply by Flooze-IT** [reply]
Useful for global attributes but i need dynamic attributes ;)

**Comment by Flooze-IT** [comment, marked as answer]
I found the solution, i need to use the `LogRecordBuilder` : 
```java
logger
            .eventBuilder(reason)
            .setBody("Error in function: "+functionName+" with this reason: "+reason)
            .setSeverity(Severity.ERROR)
            .setAttribute(AttributeKey.stringKey("requestId"),requestId)
```
and then emit it

**Comment by jkwatson** [comment]
Please note: these APIs are _not_ stable. We will be changing them, and we do not intend the logging APIs to be used by end users, only by people writing appenders for other logging frameworks. The logging APIs that we build will not be optimized for end users for that reason.
